src/charm.py

```diff
@@ -528,14 +530,61 @@ def _on_start(self, _: StartEvent) -> None:
         # would remain in maintenance status.
         self.unit.status = ActiveStatus()
 
+    def _deduped_relations(self) -> List[ops.Relation]:
+        """Return a relation list with duplicates removed.
+
+        Relations are considered duplicated if they meet the following criteria:
+        1. Two are connected via the legacy ingress and the nginx-route relation endpoint
+        2. Two share the same 4-tuple (service-hostname, service-name, service-model, service-port)
+        3. Both relations have the same remote application.
+
+        In the case of duplicates, the relation connected via legacy ingress is removed.
+
+        Returns:
+            A relation list with duplicates removed.
+        """
+
+        def key(rel: Relation) -> typing.Tuple[str, ...]:
```

Comment:
What is the reason for making this a nested function please?

Reply:
The `key` function is only used in this function, and it's the internal implementation of the `_deduped_relations` function, so it's locally scoped.

Reply:
Locally scoped functions should be rare. There are certain cases where you need local context, like for decorators. Otherwise functions shouldn't be embedded in other functions because it makes them more complicated

Reply:
Why locally scoped functions should be rare? The `key` function here is very simple and it reduce the human error of create the key separately for ingress relation and nginx-route relation. I don't see why it's complicating the code. A private function means it can be used by other methods in the class, which is not what we need.

Reply:
Because locally scoped functions make the function it is incorporated in more complicated and that locally scoped function has access to the outer scope local variables which can lead to unintended coupling between the functions and weird bugs. A private function can be used by other methods in the class, but doesn't have to be used

Reply:
This `key` function just have 5 statements, and I have already considered the variable scoping when I created this function. If you have find any incorrect variable accessing, please let me know. This coupling is intended since the `key` function is an integrating part of the `_deduped_relations` function. Isn't it better if it can not even be used by other methods in the class instead of relying on conventions?

Reply:
could call it `_gen_rel_dedup_key`

Reply:
updated, thanks